internal/ledger/compare.go

```diff
@@ -0,0 +1,413 @@
+	if record1 != nil {
+		k = stringFromBytes(record1.Key)
+		s1, err = newSnapshotRecord(record1)
+		if err != nil {
+			return nil, err
+		}
+	}
+	// Snapshot1 has a missing record
+	if record2 != nil {
+		k = stringFromBytes(record2.Key)
+		s2, err = newSnapshotRecord(record2)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return &diffRecord{
+		Namespace: namespace,
+		Key:       k,
+		Record1:   s1,
+		Record2:   s2,
+	}, nil
+}
+
+// snapshotRecord represents the data of a snapshot record in json
+type snapshotRecord struct {
+	Value    string `json:"value,omitempty"`
+	BlockNum uint64 `json:"blockNum,omitempty"`
+	TxNum    uint64 `json:"txNum,omitempty"`
+}
+
+// Creates a new SnapshotRecord
+func newSnapshotRecord(record *privacyenabledstate.SnapshotRecord) (*snapshotRecord, error) {
+	blockNum, txNum, err := heightFromBytes(record.Version)
+	if err != nil {
+		return nil, err
+	}
+
+	return &snapshotRecord{
+		Value:    stringFromBytes(record.Value),
+		BlockNum: blockNum,
+		TxNum:    txNum,
+	}, nil
+}
+
+// Obtain the block height and transaction height of a snapshot from its version bytes
+func heightFromBytes(b []byte) (uint64, uint64, error) {
+	blockNum, n1, err := util.DecodeOrderPreservingVarUint64(b)
+	if err != nil {
+		return 0, 0, err
+	}
+	txNum, _, err := util.DecodeOrderPreservingVarUint64(b[n1:])
+	if err != nil {
+		return 0, 0, err
+	}
+
+	return blockNum, txNum, nil
+}
+
+// Obtain string from bytes
+func stringFromBytes(b []byte) (str string) {
+	str = ""
+	if utf8.Valid(b) {
+		for len(b) > 0 {
+			r, size := utf8.DecodeRune(b)
+			str += string(r)
+			b = b[size:]
+		}
+	} else {
+		str = string(b)
+	}
+
+	return str
+}
+
+// nsKey is used to compare between snapshot records using both the namespace and key
+type nsKey struct {
+	namespace string
+	key       []byte
+}
+
+// Compares two nsKeys
+// Returns:
+// -1 if k1 > k2
+// 1 if k1 < k2
+// 0 if k1 == k2
+func nsKeyCompare(k1, k2 *nsKey) int {
+	res := strings.Compare(k1.namespace, k2.namespace)
+	if res != 0 {
+		return res
+	}
+	return bytes.Compare(k1.key, k2.key)
+}
+
+// Compares hashes of snapshots
+func snapshotsComparable(fpath1 string, fpath2 string) (bool, error) {
+	var mdata1, mdata2 kvledger.SnapshotSignableMetadata
+
+	// Open the first file
+	f, err := ioutil.ReadFile(fpath1)
+	if err != nil {
+		return false, err
+	}
+
+	err = json.Unmarshal([]byte(f), &mdata1)
+	if err != nil {
+		return false, err
+	}
+
+	// Open the second file
+	f, err = ioutil.ReadFile(fpath2)
+	if err != nil {
+		return false, err
+	}
+
+	err = json.Unmarshal([]byte(f), &mdata2)
+	if err != nil {
+		return false, err
+	}
+
+	if mdata1.ChannelName != mdata2.ChannelName || mdata1.LastBlockNumber != mdata2.LastBlockNumber ||
+		mdata1.LastBlockHashInHex != mdata2.LastBlockHashInHex || mdata1.StateDBType != mdata2.StateDBType {
+		return false, errors.Errorf("the supplied snapshots appear to be non-comparable. The basic details in files %s do not match",
```

Comment:
Let's indicate exactly which fields are different (e.g. if two fields don't match, report the differing values from both fields).